accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,474 @@
+- Using `ref <T>` for the strongly typed case or `ref byte` for the loosely-typed case. This is expanded in 7.0 due to the new "ref field" support. Previously, there was an internal `ByReference<T>` class that was used and in 7.0 this was changed to `ByReference` in 8.0 which no longer maintains the `<T>` type and internally just contains `ref byte`. This `ByReference` type is used today in reflection invoke when there are <=4 parameters.
+- Using unsafe `void*` (or `IntPtr`) with GC tracking or pinning to make the use GC-safe. Tracking is supported internally through the use of a newer "RegisterForGCReporting()" mechanism. This approach is used today in reflection invoke when there are >=5 parameters.
+
+# Proposed APIs
+
+## MethodInvoker
+This ref struct is the mechanism to specify the target + arguments (including return value) and supports these mechanisms:
+- `object` (including boxing). Supports loose coupling scenarios are supported, like reflection today.
+- `ref <T>`. Supports new scenarios as mentioned earlier; type must be known ahead-of-time and due to no language support, cannot be a byref-like type like `Span<T>`.
+- `void*`. Unsafe cases used to support byref-like types in an unsafe manner.
+- `TypedReference`. Optional for now; pending language asks, it may make supporting byref-like types a safe operation.
+
+```cs
+namespace System.Reflection
+{
+     public ref struct MethodInvoker
+     {
+        // Zero-arg case:
+        public MethodInvoker()
+
+        // Variable-length number of arguments:
+        public unsafe MethodInvoker(ArgumentValue* argumentStorage, int argCount)
+
+        // Fixed length (say up to 8)
+        public MethodInvoker(ref ArgumentValuesFixed values)
+
+        // Dispose needs to be called with variable-length case
+        public void Dispose()
+
+        // Target
+        public object? GetTarget()
+        public ref T GetTarget<T>()
+
+        public void SetTarget(object value)
+        public void SetTarget(TypedReference value)
+        public unsafe void SetTarget(void* value, Type type)
+        public void SetTarget<T>(ref T value)
+
+        // Arguments
+        public object? GetArgument(int index)
+        public ref T GetArgument<T>(int index)
+
+        public void SetArgument(int index, object? value)
+        public void SetArgument(int index, TypedReference value)
+        public unsafe void SetArgument(int index, void* value, Type type)
+        public void SetArgument<T>(int index, ref T value)
+
+        // Return
+        public object? GetReturn()
+        public ref T GetReturn<T>()
+
+        public void SetReturn(object value)
+        public void SetReturn(TypedReference value)
+        public unsafe void SetReturn(void* value, Type type)
+        public void SetReturn<T>(ref T value)
+
+        // Invoke direct (limited validation and defaulting)
```

Comment:
What does limited validation mean precisely?

I think it would be useful to have fully unsafe API that does not do any validation and just assumes that you have passed in arguments of the right types. Safety and performance wise, such API would be closest to Reflection.Emit where we do not perform any validation either.

Reply:
What I was thinking was to at least verify values type are the correct type to prevent destabilizing the system in a nondeterministic manner.  This can be done most efficiently in IL with castclass with minimal impact on perf. Verification on reference types should also happen, but I believe this comes for free at least.

However, the exceptions thrown will be CastClassException I believe, not a nice "ArgumentException" unless we do extra work ahead of time to throw ArgumentException or do extra work with a try\catch and back-walk the validation if CastClassException is thrown.

I was also planning on doing "defaulting" for `out` values for value types if no default values were provided by the caller. This is especially common for return values. Unfortunately, it is not easy\fast to see check if a managed reference for an argument is `in` vs `ref` vs `out` so perhaps that should be deferred for now except for the return value which is always `out`.

Reply:
I think it would be useful to have a completely unsafe API as one of the options. The completely unsafe API would not do any argument validation. It would be closest to what one can get with RefEmit where the user is responsible for emitting the right IL and where mistake can destabilize the system too.

> Verification on reference types should also happen, but I believe this comes for free at least.

Validation of reference types is not free. It is more expensive than validation of value types.

Reply:
I agree, we have the unsafe API as well.

Reply:
Updated: the `InvokeDirectUnsafe()` APIs won't do any validation.